modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+#ifndef OPENCV_CORE_QUATERNION_HPP
+#define OPENCV_CORE_QUATERNION_HPP
+
+#include <opencv2/core.hpp>
+#include <iostream>
+namespace cv
+{
+//! @addtogroup core
+//! @{
+    static constexpr double CV_QUAT_EPS = 1.e-6;
+    /**
+     * Quaternion is a number system that extends the complex numbers. It can be exprssed as a
+     * rotation in three-dimensional space.
+     * A quaternion is generally represented in the form:
+     *      \f[q = w + x\boldsymbol{i} + y\boldsymbol{j} + z\boldsymbol{k}\f]
+     *      \f[q = [w, x, y, z]\f]
+     *      \f[q = [w, \boldsymbol{v}] \f]
+     *      \f[q = ||q||[\cos\theta, u_x\sin\theta,u_y\sin\theta,  u_z\sin\theta].\f]
+     *      \f[q = ||q||[\cos\theta, \boldsymbol{u}\sin\theta]\f]
+     * where \f$\theta = \frac{\psi}{2}\f$, \f$\psi\f$ represents rotate angle,
+     * \f$\boldsymbol{u} = [u_x, u_y, u_z]\f$ represents normalized rotate axis,
+     * and \f$||q||\f$ represents the norm of \f$q\f$
+     *
+     * A unit quaternion is usually represents rotation, which has the form:
+     *      \f[q = [\cos\theta, u_x\sin\theta,u_y\sin\theta,  u_z\sin\theta].\f]
+     *
+     * To create a quaternion representing the rotation around the axis \f$\boldsymbol{v}\f$
+     * with angle \f$\psi\f$, you can use
+     * ```
+     * double angle = CV_PI;
+     * cv::Vec3d{0, 0, 1};
+     * cv::Quatd q(angle, axis);
+     * ```
+     *
+     * You can simply use four same type number to create a quaternion
+     * ```
+     * cv::Quatd q(1, 2, 3, 4);
+     * ```
+     * Or a Vec4d or Vec4f vec, then you can use
+     * ```
+     * cv::Quatd q(vec);
+     * ```
+     *
+     * If you already have a 3x3 rotate matrix R, then you can use
+     * ```
+     * cv::Quatd q(R);
+     * ```
+     *
+     * If you already have a Rodrigues vector rvec, then you can use
+     * ```
+     * cv::Quatd q(rvec);
+     * ```
+     *
+     * To extrace the rotate matrix from quaternion, see toRotMat3x3()
+     *
+     * To extrace the Vec4d or Vec4f, see toVec()
+     *
+     * To extrace the Rodrigues vector, see toRodrigues()
+     *
+     * If there are two quaternions \f$q_0, q_1\f$ interpolation is needed, you can use nlerp(), slerp() or spline()
+     * ```
+     * cv::nlerp(q0, q1, t)
+     *
+     * cv::slerp(q0, q1, t)
+     *
+     * cv::spline(q0, q0, q1, q1, t)
+     * ```
+     * spline can show the rotation smoothness between multiple quaternions
+     *
+     *
+     * Three ways to get a element in Quaternion
+     * ```
+     * Quatf q(1,2,3,4);
+     * std::cout << q.w << std::endl; // w=1,x=2,y=3,z=4
+     * std::cout << q[0] << std::endl; // q[0]=1, q[1]=2,q[2]=3,q[3]=4
+     * std::cout << q.at(0) << std::endl;
+     * ```
+     */
+    template <typename _Tp> class Quat;
+    template <typename S>
+    std::ostream& operator<<(std::ostream&, const Quat<S>&);
+
+    template <typename _Tp>
+    class Quat
+    {
+    using value_type = typename std::enable_if<
+        std::is_same<float, _Tp>::value||
```

Comment:
You can use `std::is_floating_point` and instead of SFINAE `static_assert` gives more informative output to the user with the same check:
```cpp
static_assert(std::is_floating_point<T>::value, "some informative message");
using value_type = T;
```

Reply:
I think this should ask @savuor. Can we use static_assert here ?

Reply:
`static_assert` should be OK here, see no problem.

Reply:
OK